### PR TITLE
Add verify checks to unused audio at the end

### PR DESCRIPTION
**Pull request by 64ArthurAraujo**
This PR addresses one of the missing checks listed at #12091. It compares the last placed object's position and the audio's length to calculate the percentage mapped and if it is less than 80 it returns a problem saying: 

*"There is more than 20% unused audio at the end."*

### CODE REVIEW

File: osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+        private double calculatePercentage(double mappedLenght, double trackLenght)
+        {
+            return Math.Round(mappedLenght / trackLenght * 100);
+        }
+
+        public class IssueTemplateUnusedAudioAtEnd : IssueTemplate
+        {
+            public IssueTemplateUnusedAudioAtEnd(ICheck check)
+                : base(check, IssueType.Problem, "There is more than 20% unused audio at the end.")
```
Comment (bdach) [inline osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs:42]: This issue description should probably contain the actual percentage of unmapped audio to help users gauge how much more they need to map to pass the check.

File: osu.Game/Localisation/CheckUnusedAudioAtEndStrings.cs
Comment (bdach) [inline osu.Game/Localisation/CheckUnusedAudioAtEndStrings.cs:1]: unused localisation?

File: osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+            double mappedPercentage = calculatePercentage(mappedLength, trackLength);
+
+            if (mappedPercentage < 80)
+            {
+                yield return new IssueTemplateUnusedAudioAtEnd(this).Create();
+            }
+        }
+
+        private double calculatePercentage(double mappedLenght, double trackLenght)
```
Comment (bdach) [inline osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs:34]: doesn't need to be its own method, just inline it at only point of usage

File: osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            return Math.Round(mappedLenght / trackLenght * 100);
```
Comment (bdach) [inline osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs:36]: typo in param names (should be `mappedLength`, `trackLength`)

File: osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        public CheckMetadata Metadata => new CheckMetadata(CheckCategory.Compose, "More than 20% unused audio at the end");
+
+        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
+        {
+            new IssueTemplateUnusedAudioAtEnd(this),
+        };
+
+        public IEnumerable<Issue> Run(BeatmapVerifierContext context)
+        {
+            double mappedLength = context.Beatmap.HitObjects.Last().GetEndTime();
```
Comment (bdach) [inline osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs:23]: This is not necessarily correct in gamemodes with concurrent objects (i.e. mania).

We have a separate extension for this.

```suggestion
            double mappedLength = context.Beatmap.GetLastObjectTime();
```

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
In addition to the above here, the relevant ranking criteria rule is worded thusly:

> **If you do not beatmap the last 20% of your beatmap's audio file, it should be cut.** The intro time is not included. This does not apply if more than 20% of the outro is occupied by a storyboard/video, or if more than 20% of the song's audio is not able to be mapped due to fade-out or timing issues.

So if storyboard/video is present, the issue message should include a caveat that it is fine for the tail end of audio to be occupied by it. [Mapset Verifier implements this](https://github.com/Naxesss/MapsetChecks/blob/ed77d6d0c4fd9aea41538e6aca395decce9c6637/Checks/AllModes/General/Audio/CheckAudioUsage.cs#L50-L65) so this implementation should also do so.

File: osu.Game/Localisation/CheckUnusedAudioAtEndStrings.cs
Comment (bdach) [reply osu.Game/Localisation/CheckUnusedAudioAtEndStrings.cs:1]: yes

File: osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs (hunk 1 above)
Comment (64ArthurAraujo) [reply osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs]: done

File: osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs (hunk 3 above)
Comment (64ArthurAraujo) [reply osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs]: Resolved in the same commit that made the percentage inline

File: osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs (hunk 2 above)
Comment (64ArthurAraujo) [reply osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs]: fixed

File: osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs (hunk 4 above)
Comment (64ArthurAraujo) [reply osu.Game/Rulesets/Edit/Checks/CheckUnusedAudioAtEnd.cs]: fixed

File: osu.Game/Localisation/CheckUnusedAudioAtEndStrings.cs
Comment (64ArthurAraujo) [reply osu.Game/Localisation/CheckUnusedAudioAtEndStrings.cs:1]: fixed

**Comment by bdach** [comment]
I've fixed this crashing on empty beatmaps. Looks fine otherwise.

**Review by bdach — APPROVED** [review]

**Pull request merged**